File: frontend/nfe/lib/app/data/model/nfe_detalhe_imposto_cofins_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeDetalheImpostoCofinsModel {
	int? id;
	int? idNfeDetalhe;
	String? cstCofins;
	double? baseCalculoCofins;
	double? aliquotaCofinsPercentual;
	double? quantidadeVendida;
	double? aliquotaCofinsReais;
	double? valorCofins;

	NfeDetalheImpostoCofinsModel({
		this.id,
		this.idNfeDetalhe,
		this.cstCofins,
		this.baseCalculoCofins,
		this.aliquotaCofinsPercentual,
		this.quantidadeVendida,
		this.aliquotaCofinsReais,
		this.valorCofins,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cst_cofins',
		'base_calculo_cofins',
		'aliquota_cofins_percentual',
		'quantidade_vendida',
		'aliquota_cofins_reais',
		'valor_cofins',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cst Cofins',
		'Base Calculo Cofins',
		'Aliquota Cofins Percentual',
		'Quantidade Vendida',
		'Aliquota Cofins Reais',
		'Valor Cofins',
	];

	NfeDetalheImpostoCofinsModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		cstCofins = NfeDetalheImpostoCofinsDomain.getCstCofins(jsonData['cstCofins']);
		baseCalculoCofins = jsonData['baseCalculoCofins']?.toDouble();
		aliquotaCofinsPercentual = jsonData['aliquotaCofinsPercentual']?.toDouble();
		quantidadeVendida = jsonData['quantidadeVendida']?.toDouble();
		aliquotaCofinsReais = jsonData['aliquotaCofinsReais']?.toDouble();
		valorCofins = jsonData['valorCofins']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['cstCofins'] = NfeDetalheImpostoCofinsDomain.setCstCofins(cstCofins);
		jsonData['baseCalculoCofins'] = baseCalculoCofins;
		jsonData['aliquotaCofinsPercentual'] = aliquotaCofinsPercentual;
		jsonData['quantidadeVendida'] = quantidadeVendida;
		jsonData['aliquotaCofinsReais'] = aliquotaCofinsReais;
		jsonData['valorCofins'] = valorCofins;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		cstCofins = plutoRow.cells['cstCofins']?.value != '' ? plutoRow.cells['cstCofins']?.value : 'AAA';
		baseCalculoCofins = plutoRow.cells['baseCalculoCofins']?.value?.toDouble();
		aliquotaCofinsPercentual = plutoRow.cells['aliquotaCofinsPercentual']?.value?.toDouble();
		quantidadeVendida = plutoRow.cells['quantidadeVendida']?.value?.toDouble();
		aliquotaCofinsReais = plutoRow.cells['aliquotaCofinsReais']?.value?.toDouble();
		valorCofins = plutoRow.cells['valorCofins']?.value?.toDouble();
	}	

	NfeDetalheImpostoCofinsModel clone() {
		return NfeDetalheImpostoCofinsModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			cstCofins: cstCofins,
			baseCalculoCofins: baseCalculoCofins,
			aliquotaCofinsPercentual: aliquotaCofinsPercentual,
			quantidadeVendida: quantidadeVendida,
			aliquotaCofinsReais: aliquotaCofinsReais,
			valorCofins: valorCofins,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_detalhe_imposto_cofins_model.py
from src import db


class NfeDetalheImpostoCofinsModel(db.Model):
    __tablename__ = 'nfe_detalhe_imposto_cofins'

    id = db.Column(db.Integer, primary_key=True)
    cst_cofins = db.Column(db.String(2))
    base_calculo_cofins = db.Column(db.Float)
    aliquota_cofins_percentual = db.Column(db.Float)
    quantidade_vendida = db.Column(db.Float)
    aliquota_cofins_reais = db.Column(db.Float)
    valor_cofins = db.Column(db.Float)
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.cst_cofins = data.get('cstCofins')
        self.base_calculo_cofins = data.get('baseCalculoCofins')
        self.aliquota_cofins_percentual = data.get('aliquotaCofinsPercentual')
        self.quantidade_vendida = data.get('quantidadeVendida')
        self.aliquota_cofins_reais = data.get('aliquotaCofinsReais')
        self.valor_cofins = data.get('valorCofins')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'cstCofins': self.cst_cofins,
            'baseCalculoCofins': self.base_calculo_cofins,
            'aliquotaCofinsPercentual': self.aliquota_cofins_percentual,
            'quantidadeVendida': self.quantidade_vendida,
            'aliquotaCofinsReais': self.aliquota_cofins_reais,
            'valorCofins': self.valor_cofins,
        }